### PR TITLE
[DO-2515]Implemented Person search

### PR DESCRIPTION
**Pull request by IegorSoboliev**

### CODE REVIEW

File: person/views.py [hunk 1]
```diff
@@ -23,3 +29,59 @@ class PersonViewSet(RegisterViewMixin,
...
+    def get_serializer_class(self):
+        if self.action == 'retrieve':
+            return PersonDetailSerializer
+        return super().get_serializer_class()
+
+
+class SearchPersonView(generics.ListAPIView):
+    schema = None
+    serializer_class = PersonListSerializer
+    queryset = Person.objects.filter(is_main=True)
```
Comment (roman-tiukh) [inline person/views.py:42]:
```suggestion
    queryset = Person.objects.filter(duplicate_of__isnull=True)
```

File: person/views.py [hunk 2]
```diff
@@ -23,3 +29,59 @@ class PersonViewSet(RegisterViewMixin,
...
+    # TODO: define permission classes
+    permission_classes = [AllowAny, ]
+
+    def list(self, request, *args, **kwargs):
+        last_name_coefficient = 2
+        citizenship_weight = 1
+        last_name_threshold = 0.7
+        last_name = request.query_params.get('last_name')
```
Comment (roman-tiukh) [inline person/views.py:50]: `last_name` should be required.
`first_name`, `middle_name`, `country_name` is optional, so if they are not provided then we don't need annotate

**Review by roman-tiukh — CHANGES_REQUESTED** [review]

File: person/views.py (hunk 2 above)
Comment (IegorSoboliev) [reply person/views.py:50]: As we discussed, last_name will be required on front-end

File: person/views.py [hunk 3]
```diff
@@ -23,3 +29,59 @@ class PersonViewSet(RegisterViewMixin,
...
+            return PersonDetailSerializer
+        return super().get_serializer_class()
+
+
+class SearchPersonView(generics.ListAPIView):
+    schema = None
+    serializer_class = PersonListSerializer
+    queryset = Person.objects.filter(is_main=True)
+    # TODO: define permission classes
+    permission_classes = [AllowAny, ]
```
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/views.py:44]: # TODO: define permission classes)

**Review by roman-tiukh — CHANGES_REQUESTED** [review]
validation should be on the backend.
`first_name`, `middle_name`, `country_name` is optional, so if they are not provided then we don't need to annotate

File: person/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (roman-tiukh) [reply person/views.py:44]: IsAuthenticated better than AllowAny

File: person/views.py (hunk 2 above)
Comment (IegorSoboliev) [reply person/views.py:50]: added check)

File: person/views.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (IegorSoboliev) [reply person/views.py:44]: changed

File: person/views.py [hunk 4]
```diff
@@ -23,3 +29,67 @@ class PersonViewSet(RegisterViewMixin,
...
+    # TODO: define permission classes
+    permission_classes = [IsAuthenticated, ]
+
+    def list(self, request, *args, **kwargs):
+        # verifying presence of last_name first
+        last_name = request.query_params.get('last_name')
+        if not last_name:
+            return Response(
+                {'first_name': 'This field is required'}
+            )
```
Comment (roman-tiukh) [inline person/views.py:52]:
```suggestion
            return Response(
                {'first_name': 'This field is required'}, status=400
            )
```

**Review by roman-tiukh — APPROVED** [review]

**Review by EmmaSamoilova — APPROVED** [review]

**Pull request merged**